### ISSUE TITLE
New config options

### ISSUE THREAD

**Issue by jannis-baum** [jannis-baum/Vivify#35]
> continued discussion from #34

- [x] custom client-side JS, multiple files
- [x] support multiple custom style sheets
- [x] 1-n files with globs to ignore files in directory lists
- [x] assign `name` attribute to directory items that include full name of file/directory for custom scripting & colors
- [x] option to redirect to `~`-path
- [x] document changes/features
- [x] move all config to `vivify-server` so that `viv` doesn't depend on `jq` anymore **done in #90**
  - port & timeout are trivial
  - open command can be handled by having the option to pass an argument to `vivify-server` and then open that; we can probably even use a Node-way of doing it so we don't need any program anymore

**Comment by jannis-baum** [comment]
> @tuurep Another being a hide dot-prefixed files -toggle. <- if I was the sole user of this, I'd want to do it with a keybind, something like Ctrl + H, but that has the downside of not being discoverable.


1. Interesting thoughts about the keybinding; that would require a good bit of additional client-side JS. This also seems quite customized to me, so what do you think about just offering a config setting where people can add their own additional client-side JS? That should open up lots of possibilities and keep stuff that is too custom out of main Vivify.
2. The hiding certain files however is more straight forward that I was also already thinking about: on macOS you get these pesky `.DS_Store` files everywhere (e.g. see screenshot) that I hide for every directory listing tool I have. So I was thinking of adding a config option to specify regexes for ignoring files/directories that match.

**Comment by jannis-baum** [comment]
On another note: we could "complexify" (2.) and instead of just giving the option to hide matching files/directories also give the option to assign a CSS class. This would make the custom JS easier for your use case and probably also others.

You could assign a class to items that start in `.` and then use your custom JS to toggle the `display` property of that class.

**Comment by tuurep** [comment]
1. Yeah I'm not sure what that looks like in practice, but if it means I can script some stuff per keybinding, and it's not too arcane to do, sounds just great
2. So that's a bit like a `.gitignore` but for vivify dir list? Definitely good. Also since this is a markdown viewer (lets not forget this lol) a very sensible setup could be: show only directories and markdown files :smile:

**Comment by jannis-baum** [comment]
> 1. Yeah I'm not sure what that looks like in practice, but if it means I can script some stuff per keybinding, and it's not too arcane to do, sounds just great

Yes, if we add the option to assign classes by matching regexes then it should be about 3 lines of code on your custom config! :)

> 2. So that's a bit like a `.gitignore` but for vivify dir list? Definitely good. Also since this is a markdown viewer (lets not forget this lol) a very sensible setup could be: show only directories and markdown files 😄

Yes! And yes, Vivify is a markdown viewer but it's also just a viewer; it does e.g. do code highlighting if you open non-markdown files.

**Comment by jannis-baum** [comment]
Just realized the classes by regexes will also allow people to easily do custom color highlighting by file extension, for example. I think this is good to add, seems to bring lots of customization benefits for little overhead.

**Comment by tuurep** [comment]
I'm kinda bummed out about how slowly the page refreshes after saving file. Would this be easy to change or configurable if necessary?

I have a couple related questions with that:
* Would it be possible to refresh page even without having to save?
* Sometimes I run into this: trying to open an empty file doesn't work, it needs to have like at least one character in the contents

^ Both of the above occur to me when I use Vivify as sort of a live-notetaking side preview

**Comment by jannis-baum** [comment]
Hey @tuurep I replied to you on the plugin's repo: https://github.com/jannis-baum/vivify.vim/issues/5

**Comment by tuurep** [comment]
I'd like it if `$HOME` could be substituted to `~` on the address bar. Could this be an option? :)

i.e.:

`http://localhost:31622/viewer/home/tuure/projects/vivify/README.md`

to

`http://localhost:31622/viewer/~/projects/vivify/README.md`

**Comment by jannis-baum** [comment]
In the address bar it might be a bit tedious to implement and would make us more error-prone, especially if we want to keep both options.

If it's just about the page title on the other hand we could just provide the option to have the path components use `~` instead of the components that make up `$HOME`. Would that be enough?

**Comment by jannis-baum** [comment]
As for being able to open a URL like `http://localhost:31622/viewer/~/projects/vivify/README.md`, this was supposed to have already been working but there was a bug in the regex. I just fixed it: #75

**Comment by jannis-baum** [comment]
> In the address bar it might be a bit tedious to implement and would make us more error-prone, especially if we want to keep both options.

Actually, now that I thought about it longer, we could probably just make an option that will have express always redirect to the `~`-equivalent of the URL. That should be clean & straight forward.

**Comment by jannis-baum** [comment]
Noted! With the redirect it should be good. I'll update the list above.

**Comment by jannis-baum** [comment]
I might work on this next after the other two PRs are through :)

**Comment by tuurep** [comment]
Ok, thanks!

**Comment by jannis-baum** [comment]
Hey @tuurep I am debating the feature list above and it would be great to get your opinion on something:

I will definitely implement this

> - [ ] assign name attribute to directory items that include full name of file/directory for custom scripting & colors

because it will allow users to do just about any level of customization (e.g. your `ctrl+h`) with [CSS selectors](https://www.w3schools.com/cssref/css_selectors.php) in CSS or JS targeting files with given extensions, given patterns, etc. Therefore this will also allow hiding files with CSS, e.g. like this:

```css
.dir-list-file[name$=".pdf"] {
    /* styles .pdf files, e.g. hiding them */
}
```

However, this is obviously not as straight forward as also having a list of patterns that Vivify should ignore. Do you think we should still include the list of files to ignore anyways to make it more accessible and simple for users, or only have the CSS-selector-based thing to keep the feature list minimal but fully functional?

---

As a side note, the ultimate config for me here would be to show everything that is not git-ignored but I think this would be way too much overhead. It would probably come down to having to run `git` as a subprocess or having to parse `.gitignore`s along the way down to the root plus the git-config because a global ignore file could be specified there.

So a trade-off here might be to allow users to specify ignore-files filled with globs that they can (for example) set to their global git-ignore file which for me would already remove 90% of what I really don't want Vivify to list. Do you think this is super specific to me or do you think it's something other people would want as well?

**Comment by tuurep** [comment]
I like the idea of a global pattern ignore file.

Also as a loose idea would be cool if you could choose to only show certain file types like:

* directories
* markdown
* jupyter notebook

But that's a bit offtopic

Or perhaps it's a fair point because do the selectors only allow matching on the filename or do dir list items have an attribute for the (actual) file type?

**Comment by jannis-baum** [comment]
> I like the idea of a global pattern ignore file.

Noted!

> Also as a loose idea would be cool if you could choose to only show certain file types like:
> 
> * directories
> * markdown
> * jupyter notebook
> 
> But that's a bit offtopic

I mean this can then be achieved by either configuring the global ignore file like that or by using the CSS selectors to hide everything that doesn't match.

> Or perhaps it's a fair point because do the selectors only allow matching on the filename or do dir list items have an attribute for the (actual) file type?

Hm, by actual file type you mean the something like the output from `file`?

**Comment by tuurep** [comment]
Yeah mainly I'm thinking if differentiation between dirs and files with no extension will be a problem

But looks like we know it by the CSS as:

* `li .dir-list-directory`
* `li .dir-list-file`

But based on filename pattern matching alone it's not possible to distinguish

So yeah hmm

**Comment by jannis-baum** [comment]
Yes, for the CSS selectors it's definitely differentiable. This for example

```css
.dir-list-file[name$=".pdf"] {
    /* styles .pdf files, e.g. hiding them */
}
```

would only match files even if someone names there directory ending in `.pdf`.

The globs wouldn't be able to differentiate but I would argue that that's totally fine since it's the same for `gitignore` and globs in general.

**Comment by tuurep** [comment]
Oh yeah. Well it's a good question whether the gitignore style file is just redundant.

> set to their global git-ignore file which for me would already remove 90% of what I really don't want Vivify to list.

though this makes it sound useful still. I always had .gitignore separate per project so it didn't cross my mind.

**Comment by jannis-baum** [comment]
> though this makes it sound useful still. I always had .gitignore separate per project so it didn't cross my mind.

Yeah, might be most relevant for macOS users actually. We have to deal with a collection of files than appear randomly pretty much anywhere and this is not something I want to force into all the gitignores, especially when working with other people ^^

---

So anyways, what do you think about the list in the description of the issue as it is now?

**Comment by tuurep** [comment]
Looks good, although I don't know exactly what `1-n` means in this context

**Comment by jannis-baum** [comment]
> Looks good, although I don't know exactly what `1-n` means in this context

Sorry, I mean "one to many". I want to give the option to define a string for one path or an array of strings for multiple paths. Same with the custom CSS and JS :)

**Comment by tuurep** [comment]
Hey I'll be sitting on a train tomorrow for many hours with my laptop with nothing else to do so I can definitely finish this tomorrow

**Comment by jannis-baum** [comment]
> Hey I'll be sitting on a train tomorrow for many hours with my laptop with nothing else to do so I can definitely finish this tomorrow
> 
> E: meant to comment on the PR, so I mean, to review that and test the new options

Awesome, thanks! I'm excited to release v0.2.0. I decided to make the release once that PR is merged. The other things that are open also work well as next minor releases I think.

I also added a Homebrew package (primarily for macOS users) earlier and I think (90% sure) that it wouldn't work with the current latest release because of some code-signing Apple stuff that I fixed when we changed to Node SEA. So it'll be great to have a working version out☺️
